log-manager/src/main/java/io/airlift/log/Logging.java

```diff
@@ -198,8 +201,12 @@ public Map<String, Level> getAllLevels()
         return levels.build();
     }
 
-    public void configure(LoggingConfiguration config)
+    public synchronized void configure(LoggingConfiguration config)
     {
+        if (configured) {
+            log.info("Logging already configured; ignoring new configuration.");
```

Comment:
We shouldn't be logging this at info level. It's likely a programming error, so it should be a warning if anything.

Reply:
Multiple Bootstrap apps can be composed to be re-used together, and used independently as well. It seems like a valid use-case. The problem is that each app may want its logging to be configured differently, and that Logging configuration is shared with a singleton. It could be a programming error, or not. I'm seeing many opinions, and I've opted for `info` because we don't know, at that level.

Reply:
@electrum was [arguing](https://github.com/airlift/airlift/pull/1016#discussion_r1001185166) for silently ignoring it. I feel like this is something worth logging, or someone could be very surprised down the road and waste time investigating.

Reply:
Yes, that's why I suggested it should be logged at warning level.

Reply:
You're misrepresenting what I said.

The original problem that you were trying to solve was an application where `Bootstrap` was initialized twice. I said we could solve this trivially by adding `doNotInitializeLogging()` to the second invocation. You didn't like this solution, IIRC because the fact that the first invocation initialized logging would not be obvious.

By adding this logging at warning level, you're now forced to change the application to do what I originally suggested and add `doNotInitializeLogging()` to the second invocation to avoid the warning. At that point, this change is no longer necessary, although it's still good to have as a safety check.

Reply:
I'll copy the linked comment here:

> I don't think we want to log a warning when the configuration is the same. It would be better to silently ignore it. Otherwise, this is effectively making it a mistake to initialize logging twice, with the result being that we should change any such application code to not initialize the second time. But if we were to do that, we wouldn't need this change at all (except to catch mistakes).

Reply:
Sorry, not trying to misrepresent anything you said, only what I understood of it. Thanks for clarifying!

I've updated the PR to log a warning.

Using `doNotInitializeLogging()` to suppress the warning and work around this "leaky abstraction" seems to be the best option users have to acknowledge that they known what they are doing.